### PR TITLE
ESQL: Re-enable logical dependency check

### PR DESCRIPTION
**Pull request by alex-spies**
Relates https://github.com/elastic/elasticsearch/issues/105436

The last issue to fix for the logical plan dependency check to be enabled is how we treat `Enrich`'s references. (C.f. previous attempt discussed [here](url).)

`Enrich.references()` returns both the match field _and_ references to the enrich fields. E.g. for
```
... | ENRICH languages_idx ON first_name WITH lang = language_code
```
`Enrich.references()` will return `first_name`, but also `language_code`.

The references to enrich fields (`language_code` in the example) are actually irrelevant to the dependency check, because `ENRICH` adds the enrich fields, it doesn't reference fields from its input plan.

Account for this by overriding the `references` method of `LogicalPlanDependencyCheck`, so that only the match field is considered.

### CODE REVIEW

**Review by astefan — COMMENTED** [review]
Added one comment (really minor), but also @nik9000 is right: if we change a fundamental behavior from QueryPlan it very likely means we are not approaching the problem from the right angle. I don't have a suggestion off the top of my head, but I'd double and triple check the logical approach behind references in enrich, grok and dissect.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/QueryPlan.java [hunk 1]
```diff
@@ -27,7 +27,7 @@ public abstract class QueryPlan<PlanType extends QueryPlan<PlanType>> extends No
     private AttributeSet lazyOutputSet;
     private AttributeSet lazyInputSet;
     private List<Expression> lazyExpressions;
-    private AttributeSet lazyReferences;
+    protected AttributeSet lazyReferences;
```
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/QueryPlan.java]: Yeah, messing with/changing the meaning of the references probably goes too far. I went with a more local approach instead.

**Comment by alex-spies** [comment]
Discussed with @costin:
- The problem is that `Enrich.enrichFields` is `List<NamedExpression>` - because of that, the enrich fields get picked up when running `Enrich.references()` - but they do not refer to any attributes from the input plan, so they are irrelevant to the dependency check.
-  Rather than overriding `Enrich.references`, we could try to somehow exclude the enrich fields by maybe overriding `Enrich.expressions()` or somehow messing with `Enrich.info()`.

I investigated the latter but found this to not be straight forward and am afraid of unintended side effects when changing this.

Instead, I ended up only changing `LogicalPlanDependencyCheck.references(LogicalPlan p)`; this appears to be a nice, self-contained place where it's easy to reason about why we should only consider the match field, but not the enrich fields.

**Review by costin — APPROVED** [review]
The problem with Enrich remains however this change is small enough to have it enabled.

**Pull request merged**